src/backend/distributed/transaction/distributed_transaction_management.c

```diff
@@ -0,0 +1,431 @@
+} DistributedTransactionShmemData;
+
+
+static shmem_startup_hook_type prev_shmem_startup_hook = NULL;
+static DistributedTransactionShmemData *distributedTransactionShmemData = NULL;
+static DistributedTransactionBackendData *MyDistributedTransactionBackend = NULL;
+
+
+static void DistributedTransactionManagementShmemInit(void);
+static size_t DistributedTransactionManagementShmemSize(void);
+
+
+PG_FUNCTION_INFO_V1(assign_distributed_transaction_id);
+PG_FUNCTION_INFO_V1(get_distributed_transaction_id);
+
+
+/*
+ * assign_distributed_transaction_id updates the shared memory allocated for this backend
+ * and sets initiatorNodeIdentifier, transactionId, timestamp fields with the given
+ * inputs. Also, the function sets the database id via MyDatabaseId variable provided
+ * by "misadmmin.h".
+ */
+Datum
+assign_distributed_transaction_id(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	/* MyDistributedTransactionBackend should always be avaliable, just out of paranoia */
+	if (!MyDistributedTransactionBackend)
+	{
+		ereport(ERROR, (errmsg("backend is not ready for distributed transactions")));
+	}
+
+	SpinLockAcquire(&MyDistributedTransactionBackend->mutex);
+
+	MyDistributedTransactionBackend->databaseId = MyDatabaseId;
+
+	MyDistributedTransactionBackend->transactionId.initiatorNodeIdentifier =
+		PG_GETARG_INT64(0);
+	MyDistributedTransactionBackend->transactionId.transactionId = PG_GETARG_INT64(1);
+	MyDistributedTransactionBackend->transactionId.timestamp = PG_GETARG_TIMESTAMPTZ(2);
+
+	SpinLockRelease(&MyDistributedTransactionBackend->mutex);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * get_distributed_transaction_id returns a tuple with (databaseId, initiatorNodeIdentifier,
+ * transactionId, timestamp) that exists in the shared memory associated with this backend.
+ */
+Datum
+get_distributed_transaction_id(PG_FUNCTION_ARGS)
+{
+	ReturnSetInfo *returnSetInfo = (ReturnSetInfo *) fcinfo->resultinfo;
+	TupleDesc tupleDescriptor = NULL;
+	Tuplestorestate *tupleStore = NULL;
+	MemoryContext perQueryContext = NULL;
+	MemoryContext oldContext = NULL;
+
+	int attributeCount = 4;
+	Datum values[attributeCount];
+	bool isNulls[attributeCount];
+
+	DistributedTransactionId *distributedTransctionId = NULL;
+
+	CheckCitusVersion(ERROR);
+
+	/* check to see if caller supports us returning a tuplestore */
+	if (returnSetInfo == NULL || !IsA(returnSetInfo, ReturnSetInfo))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("set-valued function called in context " \
+						"that cannot accept a set")));
+	}
+
+	if (!(returnSetInfo->allowedModes & SFRM_Materialize))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("materialize mode required, but it is not " \
+						"allowed in this context")));
+	}
+
+	/* build a tuple descriptor for our result type */
+	if (get_call_result_type(fcinfo, NULL, &tupleDescriptor) != TYPEFUNC_COMPOSITE)
```

Comment:
I don't quite understand why you'd copied the SRF code, rather than just the `get_call_result_type()` bit, which should be all that you need?